src/driver.ml

```diff
@@ -699,69 +637,53 @@ let relocate_mapper = object
+  match Ast_io.Binary_ast.read_from_input read_input with
+  | Ok { input_name = ast_input_name; input_version; ast } ->
+    let ast_kind = Intf_or_impl.kind ast in
     if not (Kind.equal kind ast_kind) then
-      Location.raise_errorf ~loc:(Location.in_file fn)
-        "File contains a binary %s AST but an %s was expected"
-        (Kind.describe ast_kind)
-        (Kind.describe kind);
-    if String.equal ast_input_name input_name || not relocate then begin
+      create_versioned_error_f input_version fn ~f:(fun create_f ->
+        create_f "File contains a binary %s AST but an %s was expected"
+        (Kind.describe ast_kind) (Kind.describe kind))
+    else if String.equal ast_input_name input_name || not relocate then (
       set_input_name ast_input_name;
-      (ast_input_name, ast)
-    end else
-      (input_name,
-       Intf_or_impl.map_with_context ast relocate_mapper
-         (ast_input_name, input_name))
-
-  | Error (Unknown_version _) ->
-    Location.raise_errorf ~loc:(Location.in_file fn)
-      "File is a binary ast for an unknown version of OCaml"
-  | Error (Not_a_binary_ast prefix_read_from_file) ->
-    (* To test if a file is an AST file, we have to read the first few bytes of the
-       file. If it is not, we have to parse these bytes and the rest of the file as
-       source code.
-
-       The compiler just does [seek_on 0] in this case, however this doesn't work when
-       the input is a pipe.
-
-       What we do instead is create a lexing buffer from the input channel and pre-fill
-       it with what we read to do the test. *)
-    let lexbuf = Lexing.from_channel ic in
-    let len = String.length prefix_read_from_file in
-    Bytes.blit_string ~src:prefix_read_from_file ~src_pos:0 ~dst:lexbuf.lex_buffer ~dst_pos:0
-      ~len;
-    lexbuf.lex_buffer_len <- len;
-    lexbuf.lex_curr_p <-
-      { pos_fname = input_name
-      ; pos_lnum  = 1
-      ; pos_bol   = 0
-      ; pos_cnum  = 0
-      };
-    Lexer.skip_hash_bang lexbuf;
-    match kind with
-    | Intf -> input_name, Intf (Parse.interface      lexbuf)
-    | Impl -> input_name, Impl (Parse.implementation lexbuf)
+      Ok (ast_input_name, input_version, ast) )
+    else
+      Ok
+        ( input_name,
+          input_version,
+          Intf_or_impl.map_with_context ast relocate_mapper
+            (ast_input_name, input_name) )
+  | Error (Unknown_version (unknown_magic, fall_back_version)) ->
+    create_versioned_error_f fall_back_version fn ~f:(fun create_f ->
+      create_f "File is a binary ast for an unknown version of OCaml with magic number '%s'" unknown_magic)
+  | Error (System_error (error, fall_back_version)) ->
+    Error (error, fall_back_version)
+  | Error (Not_a_binary_ast (prefix_read_from_file, fallback_version)) ->
+    let parse_input =
+      if String.equal fn "-" then Ast_io.Source_code.Stdin prefix_read_from_file
+      else File fn
+    in
+    match Ast_io.Source_code.parse parse_input ~kind ~input_name with
```

Comment:
One change compared to before: if the input is a source file rather than a binary AST, then we open it twice:

- once for checking whether it is a binary AST or a source file
- once to parse it

This has two implications:

1. it is slightly less efficient
2. there is a race condition; if the file is modified or deleted in between the parsing might fail

Given that this case is the most commonly used in dune builds (because dune calls the ppx driver on the source file), it seems worth optimising. The second might seem a bit far-fetched, however I have observed it in practice in the past and it was pretty difficult to debug. So it seems worth getting rid of the double file opening for this reason as well.

In conclusion, we should preserve the old behaviour and arrange the code so that the file is opened only once.

Reply:
Thanks! I didn't see that performance issue.